src/ntlm_auth_wrap.c

```diff
@@ -0,0 +1,308 @@
+	for (int i = 1; i < argc; i++) {
+		// split the argument on = and check the first part to reject excluded args.
+		if (!args.insecure)
+			if ((strncmp(argv[i], "--password", strlen("--password")) == 0) ||
+			    (strncmp(argv[i], "--challenge", strlen("--challenge")) == 0))
+				continue;
+
+		char *tmpstr = log_msg;
+		log_msg = NULL;
+		asprintf(&log_msg, "%s %s ", tmpstr, argv[i]);
+	}
+	syslog(args.level, "%s time: %g ms, status: %i, exiting pid: %i", log_msg,
+	       elapsed, WEXITSTATUS(status), ppid);
+	closelog();
+}
+
+// send to statsd 
+void send_statsd(const struct arguments args, int status, double elapsed)
+{
+	struct addrinfo *ailist;
+	struct addrinfo hint;
+	int sockfd, err;
+	memset(&hint, 0, sizeof(hint));
+	hint.ai_socktype = SOCK_DGRAM;
+	hint.ai_family = AF_INET;
+	if (args.noresolv)
+		hint.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+	hint.ai_canonname = NULL;
+	hint.ai_addr = NULL;
+	hint.ai_next = NULL;
+	if ((err = getaddrinfo(args.host, args.port, &hint, &ailist)) != 0)
+		sprintf("getaddrinfo error: %s", gai_strerror(err));
+
+	if ((sockfd = socket(ailist->ai_family, SOCK_DGRAM, 0)) < 0) {
+		err = errno;
+		fprintf(stderr, "cannot contact %s:%s: %s\n", args.host, args.port, strerror(err));
+	} else {
+		char *buf;
+		char hostname[MAX_STR_LENGTH];
+		gethostname(hostname, sizeof(hostname));
+		connect(sockfd, ailist->ai_addr, ailist->ai_addrlen);
+
+		asprintf(&buf, "%s.ntlm_auth.time:%g|ms\n", hostname, elapsed);
+
+		send(sockfd, buf, strlen(buf), 0);
+
+		// increment counter if auth failed
+		if (status > 0) {
+			asprintf(&buf, "%s.ntlm_auth.failures:1|c\n", hostname);
+			send(sockfd, buf, strlen(buf), 0);
+		}
+	}
+}
+
+int main(argc, argv, envp)
+int argc;
+char **argv, **envp;
+{
+
+	/* Default values. */
+	struct arguments arguments;
+	arguments.insecure = 0;
+	arguments.nostatsd = 0;
+	arguments.noresolv = 0;
+	arguments.log = 0;
+	arguments.binary = "/usr/bin/ntlm_auth";
+	arguments.host = "localhost";
+	arguments.port = "8125";
+	arguments.facility = LOG_LOCAL5;
+	arguments.level = LOG_INFO;
+	/* Parse our arguments; every option seen by parse_opt will
+	   be reflected in arguments. */
+	argp_parse(&argp, argc, argv, 0, 0, &arguments);
+
+	struct timeval t1, t2;
+	double elapsed;
+	gettimeofday(&t1, NULL);
+
+	// Fork a process, exec it and then wait for the exit.
+	pid_t pid, ppid;
+	ppid = getpid();
+
+	int status;
+	if ((pid = fork()) < 0) {
+		perror(argv[0]);
+		exit(1);
+	} else if (pid == 0) {	// child
+		// Find the -- separator if any and reset the argv accordingly
+		// so it does not get passed to the execed program.
+		int opt_end = 0;
+		for (int i = 1; i < argc; i++) {
+			if ((strncmp(argv[i], "--", strlen("--"))) == 0) {
+				opt_end = i;
+				break;
+			}
+		}
+		if (opt_end)
+			argv += opt_end;
+
+		argv[0] = arguments.binary;
+		execve(arguments.binary, argv, envp);
+		perror(argv[0]);
+		exit(1);
+	}
+	if (waitpid(pid, &status, 0) != pid) {	// wait for child
```

Comment:
As discussed with Louis offline we should do this until we know how it will interact with freeradius

You can see an example that without a goto here.
http://man7.org/tlpi/code/online/dist/procexec/system.c.html